Core/src/RequestWrapper.php

```diff
@@ -309,54 +311,102 @@ private function applyHeaders(RequestInterface $request, array $options = [])
+
+            foreach ($authHeaders as $key => $value) {
+                if (array_key_exists($key, $headers)) {
+                    $headers[$key] .= ' ' . $value;
+                } else {
+                    $headers[$key] = $value;
+                }
+            }
         }
 
         return Utils::modifyRequest($request, ['set_headers' => $headers]);
     }
 
     /**
-     * Fetches credentials.
+     * Fetches auth headers.
      *
      * @param FetchAuthTokenInterface $credentialsFetcher
      * @return array
      * @throws ServiceException
      */
-    private function fetchCredentials(FetchAuthTokenInterface $credentialsFetcher)
+    private function fetchAuthHeaders(FetchAuthTokenInterface $credentialsFetcher)
     {
         $backoff = new ExponentialBackoff($this->retries, $this->getRetryFunction());
 
         try {
             return $backoff->execute(
                 function () use ($credentialsFetcher) {
-                    if ($token = $credentialsFetcher->fetchAuthToken($this->authHttpHandler)) {
-                        return $token;
+                    // Proper error message capture to as to not break existing
+                    // exception message.
+                    $exceptionMessage = '';
+
+                    // We need to find the actual fetcher incase of a cache wrapper
+                    // so that we can avoid the exception case where actual fetcher
+                    // does not implements UpdateMetadataInterface with cache wrapper's
+                    // `updateMetadata` being called.
+                    $actualFetcher = $credentialsFetcher;
+                    if ($credentialsFetcher instanceof FetchAuthTokenCache) {
+                        $actualFetcher = $credentialsFetcher->getFetcher();
+                    }
+
+                    if ($actualFetcher instanceof UpdateMetadataInterface) {
+                        $header = $credentialsFetcher->updateMetadata([], null, $this->authHttpHandler);
+                        if (array_key_exists('authorization', $header) ||
+                            array_key_exists('proxy-authorization', $header)) {
+                            return $header;
+                        }
+                        $exceptionMessage = 'Unable to fetch auth headers';
+                    } else {
+                        $tokenArray = $credentialsFetcher->fetchAuthToken($this->authHttpHandler);
+                        if ($this->isValidToken($tokenArray)) {
+                            return $this->createAuthHeader($tokenArray['access_token']);
+                        }
+                        $exceptionMessage = 'Unable to fetch token';
                     }
                     // As we do not know the reason the credentials fetcher could not fetch the
                     // token, we should not retry.
-                    throw new \RuntimeException('Unable to fetch token');
+                    throw new \RuntimeException($exceptionMessage);
```

Comment:
Let's not change the error message, and keep it consistent for simplicity. Changing it could potentially break someone, especially since the common path is going to be through `UpdateMetadataInterface` (the other one is for a very rare edge case), and the change in message is not all that helpful anyway.

Reply:
Makes sense. Reverted the change of exceptions messages.